server/command.js

```diff
@@ -0,0 +1,243 @@
+$.utils.command.prepositions['beside'] = 'beside';
+$.utils.command.prepositions['for'] = 'for/about';
+$.utils.command.prepositions['about'] = 'for/about';
+$.utils.command.prepositions['is'] = 'is';
+$.utils.command.prepositions['as'] = 'as';
+$.utils.command.prepositions['off'] = 'off/off of';
+$.utils.command.prepositions['off of'] = 'off/off of';
+
+$.utils.command.prepositionsRegExp = new RegExp(
+    '(?:^|\\s)(with|using|upon|underneath|under|to|through|over|out +of|' +
+    'onto|on +top +of|on|off +of|off|is|into|inside|in +front +of|' +
+    'in|from +inside|from|for|beside|beneath|behind|at|as|about)(?:\\s|$)');
+
+$.utils.command.parse = function(command, user) {
+  // Parse a user's command into components:
+  // verb, argstr, args, dobjstr, dobj, prepstr, iobjstr, iobj
+  var args = command.split(/ +/);
+  var verb = args.shift();
+  if (!verb) {
+    return undefined;
+  }
+  var argstr = command.substring(command.indexOf(verb) + verb.length + 1);
+  var dobjstr = '';
+  var dobj = null;
+  var iobjstr = '';
+  var iobj = null;
+  var prepstr = '';
+  var m = argstr.match($.utils.command.prepositionsRegExp);
+  if (m) {
+    prepstr = m[1];
+    var i = argstr.indexOf(prepstr);
+    if (i === -1) {
+      throw RangeError('Impossible preposition: ' + prepstr);
+    }
+    dobjstr = argstr.substring(0, i - 1);
+    iobjstr = argstr.substring(i + prepstr.length + 1);
+    prepstr = prepstr.replace(/ +/g, ' ');
+  } else {
+    dobjstr = argstr;
+  }
+  if (dobjstr) {
+    var m = $.utils.match(dobjstr, user);
+    if (!m.length) {
+      dobj = undefined;
+    } else if (m.length === 1) {
+      dobj = dobj[0];
+    } else {
+      dobj = ReferenceError(dobj.length + ' matches.');
+    }
+  }
+  if (iobjstr) {
+    var m = $.utils.match(iobjstr, user);
+    if (!m.length) {
+      iobj = undefined;
+    } else if (m.length === 1) {
+      iobj = iobj[0];
+    } else {
+      iobj = ReferenceError(iobj.length + ' matches.');
+    }
+  }
+  return {
+    verb: verb,
+    argstr: argstr,
+    args: args,
+    dobjstr: dobjstr,
+    dobj: dobj,
+    prepstr: prepstr,
+    iobjstr: iobjstr,
+    iobj: iobj
+  };
+};
+
+$.utils.command.execute = function(command, user) {
+  // Parse and execute a user's command.
+  // Return true if successful, false if not.
+  var spec = $.utils.command.parse(command, user);
+  if (!spec) {
+    // TODO: Support trapping of empty commands.
+    return true;
+  }
+  // Collect all objects which could host the verb.
+  var hosts = [user, user.location, spec.dobj, spec.iobj];
+  for (var i = 0; i < hosts.length; i++) {
+    var host = hosts[i];
+    if (!host) {
+      continue;
+    }
+    // Check every verb on each object for a match.
+    var func, verb, dobj, prepstr, iobj;
+    for (var prop in host) {
+      func = host[prop];
+      if (typeof func === 'function' && (prepstr = func.prepstr) &&
+          (dobj = func.dobj) && (iobj = func.iobj && (verb = func.verb))) {
+        // This is a verb.
```

Comment:
> How about prepspec instead of prepstr (etc.)?

Str actually means something, a string, whereas spec is meaningless.  My use of spec as a local variable here was a signal that I'd given up trying to create a meaningful name for that data object.  But at least it's local.  Let's not export that to users.

>What if verb were the only required one to flag verbiness, and the other default to 'any' if absent?

These properties will be created/populated automatically by the editor, so being sparse doesn't save the users any work.  And being sparse would mean that all functions that look at verbs would need to be more complicated.  I'm only seeing negatives here.

> Bonus feature: if dobjspec or iobjspec is not a string but a $.physical object then that particular object is the only valid one for this verb (or: only children of it are valid).

Writing the web-based editor where that's readable and writable would be a challenge.  I'm totally down for having d/iobjstr be a standard '$.foo.bar' string though.  But that's something we can add later.

> So much avoidable nesting.Probably better to do the spec.verb.search(verb) === 0 check first, since that is the least likely to match.

Agreed.

> Do you mean to use == instead of ===?

Good catch.